### ISSUE TITLE
Falcoctl as a plugin/rules manager

### ISSUE THREAD

**Issue by andreabonanno** [falcosecurity/falcoctl#103]
**What would you like to be added**:
I do think that at this moment Falcoctl is a good candidate to be revamped via the addition of new features and capabilities, aimed towards Falco-related artifacts (plugins, and possibly rules): 

- retrieving artifacts metadata
- fetching and managing of artifacts versions
- automated deployment of plugins/rules
- all of the above with a UX similar with the one of a package manager

**Why is this needed**:
At the moment, there is a lack of a tool able to manage and distribute Falco artifacts (plugins, rules) with a good UX.

- Plugin artifacts are simply stored in an S3 bucket. There’s no automatic tool that can be used to retrieve and install them.
- Rules are updated only on Falco releases basis. This may be a problem, users may want to automatically download the most recent version of rules, as soon as they are published. Same goes for new versions of plugins.
- At the moment, various steps with human intervention are required to fetch and install plugins and their rules.

Given the increasing number of plugins and their probable future growth as a popular and key element of Falco, this is in my opinion an important point to address. Moreover, as Falco plugins also offer observability of external (cloud) resources from any kind of host, their adoption will be likely high under Windows/MacOS: having a tool able to manage them on these environments with a lean and simple UX is of the utmost importance.
Plugins are now tracked via the [plugin registry](https://github.com/falcosecurity/plugins/blob/master/registry.yaml), which represent the single source of truth for official plugins info and is a good starting point to track where each artifact is/will be published. It would be great to have a tool able to process this data, retrieve the correct plugin/rule artifact and manage them for the user.
A valid and successful example of artifact management are the [Helm Charts](https://helm.sh/docs/topics/charts/), which could represent a source of inspiration for distribution/tracking/management of Falco-related artifacts.

**Comment by loresuso** [comment]
Hi @andreabonanno ! I totally agree with this and I am up for revamping `falcoctl` together 😎

**Comment by leogr** [comment]
cc @alacuku @loresuso

**Comment by loresuso** [comment]
After a discussion with @leogr and @alacuku, I'd like to propose a change of direction regarding the evolution of `falcoctl`, especially for the part relevant to rules and plugins distribution. We still think that this project needs to be revamped to manage Falco artifacts, but we believe that the solution proposed above is not optimal. We did some research and we ended up discovering how other projects in the CNCF landscape are using [OCI Artifacts](https://github.com/opencontainers/artifacts) distributed via container registry that are compliant with the [OCI Distribution](https://github.com/opencontainers/distribution-spec) specification. Since many of the publicly available container registries (e.g. GHCR) implement the features requested by the specification, we can start defining new kind of artifacts for rules and plugins, store metadata regarding dependencies and more, and implement a tool able to manage them. This approach has several advantages: 

- we don't need to implement the server side of the solution. There's plenty of already compliant container registries. Our initial preferral lean towards GHCR, since we are using Github services almost everywhere in the other repos of the organization (like Github Actions, Github Pages). All we have to do is to define new `mediaType` for our own artifacts.
- plenty of libraries to push and pull generic artifacts to container registries. An interesting one is ORAS (OCI Registry As Storage). We are currently experimenting with it. Also, notice that this library can work with different types of container registry: in the future we could switch to e.g Amazon ECR and the code used to managed artifacts would still remain the same. 
- we solve the problem of distribution using the tools offered by the CNCF landscape. This directions is already taken by other CNCF projects like OPA and Helm. 
- this approach allows third parties to set up their own registry and use falcoctl in the same way. We believe that this is really important, since an organization may have worked on its own rules or plugins that may be not publicly available. 

For these reasons, you will see a bunch of PRs in the following days aiming at moving in this direction. At first, there will be a cleanup of the features that were implemented following the old approach (like the commands `repo add` for instance). Then, little by little we will introduce the commands to interact with the OCI Registry. Notice that the commands are exactly the same used e.g. by Docker to interact with a registry:

- `login` — used to login to the remote registry
- `logout` — used to logout from the remote registry
- `push` — used to store a new artifact in the registry. Artifacts will be indexed like containere images, so using a repository name and tags. The push operation will also set all the necessary metadata needed to describe the artifact and its dependencies if needed. This command will be useful in the CI when new rules or plugins needs to be published.
- `pull` — used to retrieve an artifact from the registry. Also responsible to manage dependencies while pulling. This command will be useful anytime the user needs to download artifacts. 

The OCI Distribution specification still not provide us an API to *search* repositories in a registry. We can rely on an external file containing the information we already have in the [plugin registry](https://github.com/falcosecurity/plugins/blob/master/registry.yaml) to overcome this issue. We still need to better figure where to store it, but for sure we can use it to search by keywords and retrieve a description. Then, we can conventionally say that the name of an entry there corresponds to a repository in the registry.

Falcoctl will also be useful to retrieve plugins and rules also in a k8s environment. By letting Falcoctl run as Init Container and let it consume the Falco configuration, it can download and install all the needed artifacts and make them available to Falco via an empty dir. More on that in the next steps! 

These are the overall ideas we come up with about `falcoctl`. Feel free to raise questions or concerns, we are wide open to discuss :)
